### PR TITLE
Backport collection indicies

### LINKED ISSUE #3725: [Backport] Collection indicies
### Original feature issue

- Issue: #3169
- PR: #3671

### Target release

- [x] release33
- [ ] release32
- [ ] release30

### Reason

To counter the performance issues the search has to work faster. As such the collection indices should be backported.

### PR DESCRIPTION
**Pull request by SchoolGuy**
## Linked Items

Fixes #3725

## Description

Adds indexing of collections by non-key properties.

## Behaviour changes

Old: Quick search is supported only by item name.

New: A quick search can be performed using non-key properties.

There is additional overhead for updating indexes and allocating memory for them.

## Category

This is related to a:

- [ ] Bugfix
- [x] Feature
- [ ] Packaging
- [ ] Docs
- [ ] Code Quality
- [x] Refactoring
- [ ] Miscellaneous

## Tests

- [x] Unit-Tests were created
- [ ] System-Tests were created
- [ ] Code is already covered by Unit-Tests
- [ ] Code is already covered by System-Tests
- [ ] No tests required

### CODE REVIEW

**Comment by SchoolGuy** [comment]
@tpw56j the backport is ready in my opinion. I think I successfully tested the PR locally but I'd like to be extra sure. Since you created the code originally it would be lovely if you have time in the coming days if you'd have a look over it.

File: cobbler/cobbler_collections/collection.py [hunk 1]
```diff
@@ -488,6 +487,58 @@ def _deserialize(self) -> None:
...
+
+    def remove_from_indexes(self, ref: "ITEM") -> None:
+        """
+        Remove index keys for the object.
+        :param ref: The reference to the object whose index keys are removed.
+        """
+        indx_dict = self.indexes["uid"]
+        indx_dict.pop(ref.uid, None)
+
+    def find_by_indexes(self, kargs: Dict[str, Any]) -> Optional[List["ITEM"]]:
```
Comment (m-czernek) [inline cobbler/cobbler_collections/collection.py:506]: Could we rename `kargs` to `kwargs`? In my mind, `kargs` is a bit confusing to me, it's either `args`, or `kwargs`.

I see btw that the codebase uses a lot of `kargs`, which seem interchangeable with `kwargs`. Confusingly, it also uses `kwargs`.

**Review by m-czernek — APPROVED** [review]
+1, did some basic testing, seems to work to me. Leaving a non-blocking comment, we probably won't want to modify this particular PR since it's a backport, but would like to know your opinion.

File: cobbler/cobbler_collections/collection.py (hunk 1 above)
Comment (SchoolGuy) [reply cobbler/cobbler_collections/collection.py:506]: Agreed. I'll open an issue for it.

Comment (SchoolGuy) [reply cobbler/cobbler_collections/collection.py:506]: https://github.com/cobbler/cobbler/issues/3732

**Pull request merged**